src/main/java/org/killbill/billing/client/KillBillHttpClient.java

```diff
@@ -391,87 +391,121 @@ public Response doOptions(final String uri, final Multimap<String, String> optio
 
     private String getUniqueValue(final Multimap<String, String> options, final String key) {
         final Collection<String> values = options.get(key);
-        if (values == null || values.size() == 0) {
+        if (values == null || values.isEmpty()) {
             return null;
         } else {
             Preconditions.checkState(values.size() == 1, "You can only specify a unique value for " + key);
             return values.iterator().next();
         }
     }
 
-    private <T> T executeAndWait(final BoundRequestBuilder builder, final int timeoutSec, final Class<T> clazz) throws KillBillClientException {
-        Response response;
-        final ListenableFuture<Response> futureStatus;
+    private Response request(final BoundRequestBuilder builder, final int timeoutSec) throws KillBillClientException {
         try {
-            futureStatus = builder.execute(new AsyncCompletionHandler<Response>() {
+            final ListenableFuture<Response> futureStatus = builder.execute(new AsyncCompletionHandler<Response>() {
                 @Override
                 public Response onCompleted(final Response response) throws Exception {
                     return response;
                 }
             });
-            response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+            return futureStatus.get(timeoutSec, TimeUnit.SECONDS);
         } catch (final InterruptedException e) {
             throw new KillBillClientException(e);
         } catch (final ExecutionException e) {
             throw new KillBillClientException(e);
         } catch (final TimeoutException e) {
             throw new KillBillClientException(e);
         }
+    }
 
-        if (response != null && response.getStatusCode() == 401) {
-            throw new KillBillClientException(new IllegalArgumentException("Unauthorized - did you configure your RBAC and/or tenant credentials?"),
-                                              response);
-        } else if (response != null && (response.getStatusCode() == 404 || response.getStatusCode() == 204)) {
-            // Return empty list for KillBillObjects instead of null for convenience
-            if (Iterable.class.isAssignableFrom(clazz)) {
-                for (final Constructor constructor : clazz.getConstructors()) {
-                    if (constructor.getParameterTypes().length == 0) {
-                        try {
-                            return (T) constructor.newInstance();
-                        } catch (InstantiationException e) {
-                            return null;
-                        } catch (IllegalAccessException e) {
-                            return null;
-                        } catch (InvocationTargetException e) {
-                            return null;
-                        }
-                    }
-                }
-                return null;
+    private void throwExceptionOnResponseError(final Response response) throws KillBillClientException {
```

Comment:
This method checks the response and throws an exception in case of an error. If it doesn't throw an exception it means that the request was successful and we should continue to process it, which will usually means deserializing the response:
https://github.com/andrenpaes/killbill-client-java/blob/26e90d2b9ab03442dc4a481132e5b12dd693e6f6/src/main/java/org/killbill/billing/client/KillBillHttpClient.java#L407

What makes things complicate is that we treat responses differently depending on the `followLocation` flag.

If it's `false`, for things like 502, 503 and even 402, we should throw an exception.

If it's set to `true`, we follow the Location header if it exists, otherwise, we still want to throw an exception. Also, in the case of successful request without the Location header, we don't want to deserialize the response when `followLocation` is true, we just return null (this is based on the old behavior of the code).